x/merkledb/history.go

```diff
@@ -88,39 +85,63 @@ func (th *trieHistory) getValueChanges(startRoot, endRoot ids.ID, start []byte,
 		return newChangeSummary(maxLength), nil
 	}
 
-	// Confirm there's a change resulting in [startRoot] before
-	// a change resulting in [endRoot] in the history.
-	// [lastEndRootChange] is the last change in the history resulting in [endRoot].
-	lastEndRootChange, ok := th.lastChanges[endRoot]
+	// [endRootChanges] is the last change in the history resulting in [endRoot].
+	endRootChanges, ok := th.lastChanges[endRoot]
 	if !ok {
 		return nil, ErrRootIDNotPresent
 	}
 
-	// [startRootChanges] is the last appearance of [startRoot]
+	// Confirm there's a change resulting in [startRoot] before
+	// a change resulting in [endRoot] in the history.
+	// [startRootChanges] is the last appearance of [startRoot].
 	startRootChanges, ok := th.lastChanges[startRoot]
 	if !ok {
 		return nil, ErrStartRootNotFound
 	}
 
-	// startRootChanges is after the lastEndRootChange, but that is just the latest appearance of start root
-	// there may be an earlier entry, so attempt to find an entry that comes before lastEndRootChange
-	if startRootChanges.index > lastEndRootChange.index {
-		th.history.DescendLessOrEqual(
-			lastEndRootChange,
-			func(item *changeSummaryAndIndex) bool {
-				if item == lastEndRootChange {
-					return true // Skip first iteration
-				}
-				if item.rootID == startRoot {
-					startRootChanges = item
-					return false
-				}
-				return true
-			},
-		)
-		// There's no change resulting in [startRoot] before the latest change resulting in [endRoot].
-		if startRootChanges.index > lastEndRootChange.index {
-			return nil, ErrStartRootNotFound
+	var (
```

Comment:
should these be calculated after figuring out the insert number of the start root before the endroot?

Reply:
the end root index has to be calculated here but I moved the start root index calculation below after we know `startRootChanges` is correct

Reply:
I mean technically we could move this:

```go
	var (
		// The insert number of the last element in [th.history].
		mostRecentChangeInsertNumber = th.nextInsertNumber - 1

		// The index within [th.history] of its last element.
		mostRecentChangeIndex = th.history.Len() - 1

		// The difference between the last index in [th.history] and the index of [endRootChanges].
		endToMostRecentOffset = int(mostRecentChangeInsertNumber - endRootChanges.insertNumber)

		// The index in [th.history] of the latest change resulting in [endRoot].
		endRootIndex = mostRecentChangeIndex - endToMostRecentOffset
	)
```

into the

```go
if startRootChanges.insertNumber > endRootChanges.insertNumber {
```

clause and then recalculate the `endRootIndex` by using the `startRootIndex` later but I think it's cleaner as is